File: Samples/MapScenariosObjC/MapScenariosObjC/Scenarios/LocationSharingViewController.m
//
//  LocationSharingViewController.m
//  MapScenariosObjC
//
//  Created on 3/13/18.
//  Copyright © 2018 Phunware. All rights reserved.
//

#import <PWMapKit/PWMapKit.h>
#import <PWCore/PWCore.h>
#import <Foundation/Foundation.h>

#import "LocationSharingViewController.h"

static NSString * const didUpdateAnnotationNotificationName = @"didUpdateAnnotation";
static NSString * const deviceDisplayNameKey = @"DeviceDisplayNameKey";
static NSString * const deviceTypeKey = @"DeviceTypeKey";

#pragma mark - NSNumber Random Value Category

@implementation NSNumber (Random)

+ (CGFloat)randomCGFloat {
    return (CGFloat)arc4random() / (CGFloat)UINT32_MAX;
}

@end

#pragma mark - SharedLocationAnnotation

@interface SharedLocationAnnotation : MKPointAnnotation

@property (nonatomic, strong) PWSharedLocation *sharedLocation;

- (id)initWithSharedLocation:(PWSharedLocation *)sharedLocation;

@end

@implementation SharedLocationAnnotation

- (id)initWithSharedLocation:(PWSharedLocation *)sharedLocation {
    if (self = [super init]) {
        _sharedLocation = sharedLocation;
    }
    return self;
}

@end

#pragma mark - SharedLocationAnnotationView

@interface SharedLocationAnnotationView : MKAnnotationView

@property (nonatomic, strong) UILabel *floatingTextLabel;

- (id)initWithAnnotation:(id<MKAnnotation>)annotation reuseIdentifier:(NSString *)reuseIdentifier color:(UIColor *)color;

@end

@implementation SharedLocationAnnotationView

- (id)initWithAnnotation:(id<MKAnnotation>)annotation reuseIdentifier:(NSString *)reuseIdentifier color:(UIColor *)color {
    if (self = [super initWithAnnotation:annotation reuseIdentifier:reuseIdentifier]) {
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didUpdateAnnotation) name:didUpdateAnnotationNotificationName object:nil];
        
        [self configureFloatingTextLabel];
        
        self.image = [self circleImageWithColor:color height:15.0];
    }
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)configureFloatingTextLabel {
    self.floatingTextLabel = [UILabel new];
    
    [self addSubview:self.floatingTextLabel];
    self.floatingTextLabel.translatesAutoresizingMaskIntoConstraints = NO;
    
    [[self.floatingTextLabel.centerXAnchor constraintEqualToAnchor:self.centerXAnchor] setActive:YES];
    [[self.floatingTextLabel.topAnchor constraintEqualToAnchor:self.bottomAnchor constant:-5.0] setActive:YES];
}

- (UIImage *)circleImageWithColor:(UIColor *)color height:(CGFloat)height {
    UIGraphicsBeginImageContextWithOptions(CGSizeMake(height, height), NO, 0);
    UIBezierPath *fillPath = [UIBezierPath bezierPathWithOvalInRect:CGRectMake(0, 0, height, height)];
    [color setFill];
    [fillPath fill];
    
    UIImage *dotImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return dotImage;
}

- (void)didUpdateAnnotation {
    dispatch_async(dispatch_get_main_queue(), ^{
        __weak typeof(self) weakSelf = self;
        NSString *displayName = ((SharedLocationAnnotation *)weakSelf.annotation).sharedLocation.displayName;
        NSString *userType = ((SharedLocationAnnotation *)weakSelf.annotation).sharedLocation.userType;
        weakSelf.floatingTextLabel.text = [NSString stringWithFormat:@"%@ (%@)", displayName, userType];
    });
}

@end

#pragma mark - LocationSharingViewController

@interface LocationSharingViewController () <PWMapViewDelegate, PWLocationSharingDelegate, CLLocationManagerDelegate>

@property (nonatomic, strong) PWMapView *mapView;
@property (nonatomic, strong) CLLocationManager *locationManager;
@property (nonatomic, assign) BOOL firstLocationAcquired;

@property (nonatomic, strong) NSSet<PWSharedLocation *> *sharedLocations;
@property (nonatomic, strong) NSMutableDictionary<NSString *, SharedLocationAnnotation *> *sharedLocationAnnotations;

@property (nonatomic, strong) NSMutableDictionary<NSString *, UIColor *> *annotationColors;

@end

@implementation LocationSharingViewController

- (id)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        _buildingIdentifier = 0; // Enter your building identifier here, found on the building's Edit page on Maas portal
        // Enter your application identifier, access key, and signature key, found on Maas portal under Account > Apps
        _applicationId = @"";
        _accessKey = @"";
        _signatureKey = @"";
        
        _sharedLocationAnnotations = [NSMutableDictionary new];
        _annotationColors = [NSMutableDictionary new];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"Location Sharing";
    
    if (self.applicationId.length > 0 && self.accessKey.length > 0 && self.signatureKey.length > 0 && self.buildingIdentifier != 0) {
        [PWCore setApplicationID:self.applicationId accessKey:self.accessKey signatureKey:self.signatureKey];
    } else {
        [NSException raise:@"MissingConfiguration" format:@"applicationId, accessKey, signatureKey, and buildingIdentifier must be set"];
    }
    
    self.locationManager = [CLLocationManager new];
    self.locationManager.delegate = self;
    [self.locationManager requestWhenInUseAuthorization];
    
    self.mapView = [PWMapView new];
    self.mapView.delegate = self;
    self.mapView.locationSharingDelegate = self;
    self.mapView.sharedLocationDisplayName = [self deviceDisplayName];
    self.mapView.sharedLocationUserType = [self deviceType];
    [self.view addSubview:self.mapView];
    [self configureMapViewConstraints];
    
    [PWBuilding buildingWithIdentifier:self.buildingIdentifier completion:^(PWBuilding *building, NSError *error) {
        __weak typeof(self) weakSelf = self;
        [weakSelf.mapView setBuilding:building animated:YES onCompletion:^(NSError *error) {
            [weakSelf startManagedLocationManager];
        }];
    }];
}

- (void)startManagedLocationManager {
    PWManagedLocationManager *managedLocationManager = [[PWManagedLocationManager alloc] initWithBuildingId:self.buildingIdentifier];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.mapView registerLocationManager:managedLocationManager];
        [self.mapView startSharingUserLocation];
        [self.mapView startRetrievingSharedLocations];
        
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Settings" style:UIBarButtonItemStylePlain target:self action:@selector(settingsTapped)];
    });
}

- (void)configureMapViewConstraints {
    self.mapView.translatesAutoresizingMaskIntoConstraints = NO;
    [[self.mapView.topAnchor constraintEqualToAnchor:self.view.topAnchor] setActive:YES];
    [[self.mapView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor] setActive:YES];
    [[self.mapView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor] setActive:YES];
    [[self.mapView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor] setActive:YES];
}

- (SharedLocationAnnotationView *)sharedLocationAnnotationView:(SharedLocationAnnotation *)sharedLocationAnnotation mapView:(MKMapView *)mapView {
    SharedLocationAnnotationView *dotView = (SharedLocationAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:sharedLocationAnnotation.sharedLocation.deviceId];
    
    if (dotView == nil) {
        UIColor *color = [UIColor blackColor];
        UIColor *oldColor = self.annotationColors[sharedLocationAnnotation.sharedLocation.deviceId];
        if (oldColor) {
            color = oldColor;
        } else {
            color = [UIColor colorWithRed:[NSNumber randomCGFloat] green:[NSNumber randomCGFloat] blue:[NSNumber randomCGFloat] alpha:1.0];
            self.annotationColors[sharedLocationAnnotation.sharedLocation.deviceId] = color;
        }
        
        dotView = [[SharedLocationAnnotationView alloc] initWithAnnotation:sharedLocationAnnotation reuseIdentifier:sharedLocationAnnotation.sharedLocation.deviceId color:color];
        dotView.floatingTextLabel.text = sharedLocationAnnotation.title;
    }
    
    return dotView;
}

#pragma mark Setters and Getters

- (void)setDeviceDisplayName:(NSString *)deviceDisplayName {
    self.mapView.sharedLocationDisplayName = deviceDisplayName;
    [[NSUserDefaults standardUserDefaults] setObject:deviceDisplayName forKey:deviceDisplayNameKey];
}

- (NSString *)deviceDisplayName {
    NSString *deviceDisplayName = [[NSUserDefaults standardUserDefaults] objectForKey:deviceDisplayNameKey];
    if (deviceDisplayName) {
        return deviceDisplayName;
    }
    return @"Test";
}

- (void)setDeviceType:(NSString *)deviceType {
    self.mapView.sharedLocationUserType = deviceType;
    [[NSUserDefaults standardUserDefaults] setObject:deviceType forKey:deviceTypeKey];
}

- (NSString *)deviceType {
    NSString *deviceType = [[NSUserDefaults standardUserDefaults] objectForKey:deviceTypeKey];
    if (deviceType) {
        return deviceType;
    }
    return @"Employee";
}

#pragma mark PWMapViewDelegate

- (void)mapView:(PWMapView *)mapView locationManager:(id<PWLocationManager>)locationManager didUpdateIndoorUserLocation:(PWUserLocation *)userLocation {
    if (!self.firstLocationAcquired) {
        self.firstLocationAcquired = YES;
        self.mapView.trackingMode = PWTrackingModeFollow;
    }
}

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id<MKAnnotation>)annotation {
    return [self sharedLocationAnnotationView:(SharedLocationAnnotation *)annotation mapView:mapView];
}

#pragma mark PWLocationSharingDelegate

- (void)didUpdateSharedLocations:(NSSet<PWSharedLocation *> *)sharedLocations {
    self.sharedLocations = sharedLocations;
    
    for (PWSharedLocation *updatedSharedLocation in sharedLocations) {
        SharedLocationAnnotation *annotation = self.sharedLocationAnnotations[updatedSharedLocation.deviceId];
        if (annotation != nil) {
            annotation.sharedLocation = updatedSharedLocation;
            dispatch_async(dispatch_get_main_queue(), ^{
                [UIView animateWithDuration:0.5 animations:^{
                    annotation.title = [NSString stringWithFormat:@"%@ (%@)", updatedSharedLocation.displayName, updatedSharedLocation.userType];
                    annotation.coordinate = updatedSharedLocation.location;
                }];
            });
        }
    }
    
    [[NSNotificationCenter defaultCenter] postNotificationName:didUpdateAnnotationNotificationName object:nil];
}

- (void)didAddSharedLocations:(NSSet<PWSharedLocation *> *)addedSharedLocations {
    for (PWSharedLocation *addedSharedLocation in addedSharedLocations) {
        dispatch_async(dispatch_get_main_queue(), ^{
            SharedLocationAnnotation *annotation = [[SharedLocationAnnotation alloc] initWithSharedLocation:addedSharedLocation];
            annotation.title = [NSString stringWithFormat:@"%@ (%@)", addedSharedLocation.displayName, addedSharedLocation.userType];
            annotation.coordinate = addedSharedLocation.location;
            
            __weak typeof(self) weakSelf = self;
            weakSelf.sharedLocationAnnotations[addedSharedLocation.deviceId] = annotation;
            [weakSelf.mapView addAnnotation:annotation];
        });
    }
}

- (void)didRemoveSharedLocations:(NSSet<PWSharedLocation *> *)removedSharedLocations {
    for (PWSharedLocation *removedSharedLocation in removedSharedLocations) {
        dispatch_async(dispatch_get_main_queue(), ^{
            __weak typeof(self) weakSelf = self;
            SharedLocationAnnotation *annotation = weakSelf.sharedLocationAnnotations[removedSharedLocation.deviceId];
            if (annotation) {
                [weakSelf.mapView removeAnnotation:annotation];
                [weakSelf.sharedLocationAnnotations removeObjectForKey:removedSharedLocation.deviceId];
            }
        });
    }
}

#pragma mark - CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    if (status == kCLAuthorizationStatusAuthorizedAlways || status == kCLAuthorizationStatusAuthorizedWhenInUse) {
        [self startManagedLocationManager];
    } else {
        NSLog(@"Not authorized to start PWManagedLocationManager");
    }
}

#pragma mark Settings Button

- (void)settingsTapped {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Set device name and type" message:nil preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *okayAction = [UIAlertAction actionWithTitle:@"Okay" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        __weak typeof(self) weakSelf = self;
        weakSelf.deviceDisplayName = alertController.textFields[0].text;
        weakSelf.deviceType = alertController.textFields[1].text;
    }];
    [alertController addAction:okayAction];
    
    [alertController addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        __weak typeof(self) weakSelf = self;
        textField.text = weakSelf.deviceDisplayName;
    }];
    [alertController addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        __weak typeof(self) weakSelf = self;
        textField.text = weakSelf.deviceType;
    }];
    
    [self presentViewController:alertController animated:YES completion:nil];
}

@end

File: Samples/MapScenarios/MapScenarios/Scenarios/LocationSharingViewController.swift
//
//  LocationSharingViewController.swift
//  MapScenarios
//
//  Created on 3/12/18.
//  Copyright © 2018 Phunware. All rights reserved.
//

import Foundation
import UIKit
import PWMapKit
import PWCore

// MARK: - CGFloat extension
extension CGFloat {
    
    static func random() -> CGFloat {
        return CGFloat(arc4random()) / CGFloat(UInt32.max)
    }
}

// MARK: - SharedLocationAnnotation
class SharedLocationAnnotation: MKPointAnnotation {
    
    var sharedLocation: PWSharedLocation!
    
    init(sharedLocation: PWSharedLocation) {
        super.init()
        self.sharedLocation = sharedLocation
    }
}

// MARK: - Notification
extension Notification.Name {
    static let didUpdateAnnotation = Notification.Name("didUpdateAnnotation")
}

// MARK: - SharedLocationAnnotationView
class SharedLocationAnnotationView: MKAnnotationView {
    
    let floatingTextLabel = UILabel()
    
    init(annotation: MKAnnotation?, reuseIdentifier: String?, color: UIColor) {
        super.init(annotation: annotation, reuseIdentifier: reuseIdentifier)
        NotificationCenter.default.addObserver(self, selector: #selector(didUpdateAnnotation), name: .didUpdateAnnotation, object: nil)
        
        configureFloatingTextLabel()
        
        image = circleImageWithColor(color: color, height: 15.0)
    }
    
    func circleImageWithColor(color: UIColor, height: CGFloat) -> UIImage {
        UIGraphicsBeginImageContextWithOptions(CGSize(width: height, height: height), false, 0)
        let fillPath = UIBezierPath(ovalIn: CGRect(x: 0, y: 0, width: height, height: height))
        color.setFill()
        fillPath.fill()
        
        let dotImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return dotImage!
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}

// MARK: Floating Text Label
extension SharedLocationAnnotationView {
    
    func configureFloatingTextLabel() {
        addSubview(floatingTextLabel)
        floatingTextLabel.translatesAutoresizingMaskIntoConstraints = false
    }
    
    func addFloatingText(_ text: String) {
        floatingTextLabel.text = text
        floatingTextLabel.sizeToFit()
        
        floatingTextLabel.centerXAnchor.constraint(equalTo: self.centerXAnchor, constant: 0).isActive = true
        floatingTextLabel.topAnchor.constraint(equalTo: self.bottomAnchor, constant: -5.0).isActive = true
    }
    
    @objc func didUpdateAnnotation() {
        DispatchQueue.main.async { [weak self] in
            guard let annotation = self?.annotation as? SharedLocationAnnotation else {
                return
            }
            
            self?.addFloatingText("\(annotation.sharedLocation.displayName!) (\(annotation.sharedLocation.userType!))")
        }
    }
}

// MARK: - LocationSharingViewController
class LocationSharingViewController: UIViewController, ScenarioProtocol {
    @IBOutlet private var floorSwitchContainerView: UIView!
    @IBOutlet private var floorSwitchPickerView: UIPickerView!
    
    private var firstFloorSwitch = false
    private var floorSwitchFloors = [PWFloor]()
    private var floorSwitchBuildings = [PWBuilding]()

    // Enter your application identifier, access key, and signature key, found on Maas portal under Account > Apps
    var applicationId = ""
    var accessKey = ""
    var signatureKey = ""
    
    // Enter your campus identifier here, found on the campus's Edit page on Maas portal
    var campusIdentifier = 0

    // Enter your building identifier here, found on the building's Edit page on Maas portal
    var buildingIdentifier = 0
    
    private let mapView = PWMapView()
    private let locationManager = CLLocationManager()
    private var firstLocationAcquired = false
    
    private let deviceDisplayNameKey = "DeviceDisplayNameKey"
    private var deviceDisplayName: String {
        get {
            if let displayName = UserDefaults.standard.object(forKey: deviceDisplayNameKey) as? String {
                return displayName
            }
            return "Test"
        }
        set {
            mapView.sharedLocationDisplayName = newValue
            UserDefaults.standard.set(newValue, forKey: deviceDisplayNameKey)
        }
    }
    
    private let deviceTypeKey = "DeviceTypeKey"
    private var deviceType: String {
        get {
            if let type = UserDefaults.standard.object(forKey: deviceTypeKey) as? String {
                return type
            }
            return "Employee"
        }
        set {
            mapView.sharedLocationUserType = newValue
            UserDefaults.standard.set(newValue, forKey: deviceTypeKey)
        }
    }
    
    private var sharedLocations = Set<PWSharedLocation>()
    private var sharedLocationAnnotations = [String : SharedLocationAnnotation]()
    
    private var annotationColors = [String : UIColor]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = "Location Sharing"
        
        if !validateScenarioSettings() {
            return
        }
        
        PWCore.setApplicationID(applicationId, accessKey: accessKey, signatureKey: signatureKey)
        mapView.delegate = self
        mapView.locationSharingDelegate = self
        mapView.sharedLocationDisplayName = deviceDisplayName
        mapView.sharedLocationUserType = deviceType
        view.addSubview(mapView)
        configureMapViewConstraints()
        
        configureFloorSwitchButton()
        configureFloorSwitchContainerView()

        // If we want to route between buildings on a campus, then we use PWCampus.campus to configure MapView
        // Otherwise, we will use PWBuilding.building to route between floors in a single building.
        if campusIdentifier != 0 {
            PWCampus.campus(identifier: campusIdentifier) { [weak self] (campus, error) in
                if let error = error {
                    self?.warning(error.localizedDescription)
                    return
                }

                self?.mapView.setCampus(campus, animated: true, onCompletion: { (error) in
                    self?.locationManager.delegate = self
                    if !CLLocationManager.isAuthorized() {
                        self?.locationManager.requestWhenInUseAuthorization()
                    } else {
                        self?.startManagedLocationManager()
                    }
                })
            }
        }
        else {
            PWBuilding.building(withIdentifier: buildingIdentifier) { [weak self] (building, error) in
                if let error = error {
                    self?.warning(error.localizedDescription)
                    return
                }

                self?.mapView.setBuilding(building, animated: true, onCompletion: { (error) in
                    self?.locationManager.delegate = self
                    if !CLLocationManager.isAuthorized() {
                        self?.locationManager.requestWhenInUseAuthorization()
                    } else {
                        self?.startManagedLocationManager()
                    }
                })
            }
        }
    }
    
    func startManagedLocationManager() {
        // In order to route between buildings on a campus, we also need to register the
        // PWManagedLocationManager using campusIdentifier.  Otherwise, we will register
        // using buildingIdentifier.
        if campusIdentifier != 0 {
            DispatchQueue.main.async { [weak self] in
                guard let campusIdentifier = self?.campusIdentifier else {
                    return
                }
                let managedLocationManager = PWManagedLocationManager(campusId: campusIdentifier)
                self?.mapView.register(managedLocationManager)
                self?.mapView.startSharingUserLocation()
                self?.mapView.startRetrievingSharedLocations()
                
                if let strongSelf = self {
                    strongSelf.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Settings", style: .plain, target: strongSelf, action: #selector(strongSelf.settingsTapped))
                }
            }
        }
        else {
            DispatchQueue.main.async { [weak self] in
                guard let buildingIdentifier = self?.buildingIdentifier else {
                    return
                }
                let managedLocationManager = PWManagedLocationManager(buildingId: buildingIdentifier)
                self?.mapView.register(managedLocationManager)
                self?.mapView.startSharingUserLocation()
                self?.mapView.startRetrievingSharedLocations()
                
                if let strongSelf = self {
                    strongSelf.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Settings", style: .plain, target: strongSelf, action: #selector(strongSelf.settingsTapped))
                }
            }
        }
    }
    
    func configureMapViewConstraints() {
        mapView.translatesAutoresizingMaskIntoConstraints = false
        mapView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        mapView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
        mapView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
        mapView.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
    }
    
    func sharedLocationAnnotationView(sharedLocationAnnotation: SharedLocationAnnotation, mapView: MKMapView) -> SharedLocationAnnotationView {
        var dotView = mapView.dequeueReusableAnnotationView(withIdentifier: sharedLocationAnnotation.sharedLocation.deviceId) as? SharedLocationAnnotationView
        
        if dotView == nil {
            var color = UIColor.black
            if let oldColor = annotationColors[sharedLocationAnnotation.sharedLocation.deviceId] {
                color = oldColor
            } else {
                let newColor = UIColor(red: .random(), green: .random(), blue: .random(), alpha: 1.0)
                color = newColor
                annotationColors[sharedLocationAnnotation.sharedLocation.deviceId] = color
            }
            
            dotView = SharedLocationAnnotationView(annotation: sharedLocationAnnotation, reuseIdentifier: sharedLocationAnnotation.sharedLocation.deviceId, color: color)
            if let title = sharedLocationAnnotation.title {
                dotView!.addFloatingText(title)
            }
        }
        
        return dotView!
    }
}

// MARK: - PWMapViewDelegate
extension LocationSharingViewController: PWMapViewDelegate {
    
    func mapView(_ mapView: PWMapView!, locationManager: PWLocationManager!, didUpdateIndoorUserLocation userLocation: PWUserLocation!) {
        if !firstLocationAcquired {
            firstLocationAcquired = true
            mapView.trackingMode = .follow
        }
    }
    
    func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
        var annotationView = MKAnnotationView()
        
        if let annotation = annotation as? SharedLocationAnnotation {
            annotationView = sharedLocationAnnotationView(sharedLocationAnnotation: annotation, mapView: mapView)
        }
        
        return annotationView
    }
    
    func mapView(_ mapView: PWMapView!, didFailToLocateIndoorUserWithError error: Error!) {
        guard let error = error as NSError? else {
            return
        }
        
        let title = error.domain
        let description = error.userInfo["message"] as? String ?? "Unknown Error"
        let message = "\(description) \n Error Code: \(error.code)"
        
        showAlertForIndoorLocationFailure(withTitle: title , failureMessage: message)
    }
    
    func mapView(_ mapView: PWMapView!, didChange floor: PWFloor!) {
        guard firstFloorSwitch else {
            firstFloorSwitch = true
            return
        }
        
        mapView.zoomToFitFloor(floor)
    }
}

// MARK: - PWLocationSharingDelegate
extension LocationSharingViewController: PWLocationSharingDelegate {
    
    // Called when the shared locations are updated. Includes complete list of shared locations
    
    func didUpdate(_ sharedLocations: Set<PWSharedLocation>!) {
        self.sharedLocations = sharedLocations
        
        for updatedSharedLocation in sharedLocations {
            if let annotation = sharedLocationAnnotations[updatedSharedLocation.deviceId] {
                annotation.sharedLocation = updatedSharedLocation
                DispatchQueue.main.async {
                    UIView.animate(withDuration: 0.5, animations: {
                        annotation.title = "\(updatedSharedLocation.displayName!) (\(updatedSharedLocation.userType!))"
                        annotation.coordinate = updatedSharedLocation.location
                    })
                }
            }
        }
        
        NotificationCenter.default.post(name: .didUpdateAnnotation, object: nil)
    }
    
    // Called when a new shared location is added
    
    func didAdd(_ addedSharedLocations: Set<PWSharedLocation>!) {
        for addedSharedLocation in addedSharedLocations {
            DispatchQueue.main.async { [weak self] in
                let annotation = SharedLocationAnnotation(sharedLocation: addedSharedLocation)
                annotation.title = "\(addedSharedLocation.displayName!) (\(addedSharedLocation.userType!))"
                annotation.coordinate = addedSharedLocation.location
                
                self?.sharedLocationAnnotations[addedSharedLocation.deviceId] = annotation
                self?.mapView.addAnnotation(annotation)
            }
        }
    }
    
    // Called when a shared location is removed
    
    func didRemove(_ removedSharedLocations: Set<PWSharedLocation>!) {
        for removedSharedLocation in removedSharedLocations {
            DispatchQueue.main.async { [weak self] in
                if let annotation = self?.sharedLocationAnnotations[removedSharedLocation.deviceId] {
                    self?.mapView.removeAnnotation(annotation)
                    self?.sharedLocationAnnotations.removeValue(forKey: removedSharedLocation.deviceId)
                }
            }
        }
    }
}

// MARK: - CLLocationManagerDelegate
extension LocationSharingViewController: CLLocationManagerDelegate {
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        switch status {
        case .authorizedAlways, .authorizedWhenInUse:
            startManagedLocationManager()
        default:
            mapView.unregisterLocationManager()
            print("Not authorized to start PWLocationManager")
        }
    }
}

// MARK: - Settings Button
extension LocationSharingViewController {
    
    @objc func settingsTapped() {
        let alertController = UIAlertController(title: "Set device name and type", message: nil, preferredStyle: .alert)
        
        alertController.addAction(UIAlertAction(title: "Okay", style: .default, handler: { [weak self] (alertAction) in
            if let displayName = alertController.textFields?[0].text {
                self?.deviceDisplayName = displayName
            }
            if let type = alertController.textFields?[1].text {
                self?.deviceType = type
            }
        }))
        
        alertController.addTextField { [weak self] (textField) in
            textField.text = self?.deviceDisplayName
        }
        alertController.addTextField { [weak self] (textField) in
            textField.text = self?.deviceType
        }
        
        present(alertController, animated: true, completion: nil)
    }
}


// MARK: - User Actions

extension LocationSharingViewController {
    func configureFloorSwitchContainerView() {
        view.bringSubviewToFront(floorSwitchContainerView)
        floorSwitchContainerView.isHidden = true
    }
    
    func configureFloorSwitchButton() {
        let floorImage = UIImage(named: "Floors")
        let floorButton = UIButton()
        floorButton.setImage(floorImage, for: .normal)
        floorButton.addTarget(self, action: #selector(floorButtonTapped), for: .touchUpInside)
        floorButton.frame = CGRect(x: view.bounds.maxX - 50, y: 0, width: 50, height: 50)
        floorButton.autoresizingMask = [.flexibleLeftMargin, .flexibleBottomMargin]
        self.view.addSubview(floorButton)
    }

    @objc func floorButtonTapped(_ sender: UIBarButtonItem) {
        floorSwitchBuildings.removeAll()
        floorSwitchFloors.removeAll()
        
        guard let floors = mapView.floors, !floors.isEmpty else {
            floorSwitchContainerView.isHidden = true
            floorSwitchPickerView.reloadAllComponents()
            return
        }
        
        guard floorSwitchContainerView.isHidden else {
            floorSwitchContainerView.isHidden = true
            floorSwitchPickerView.reloadAllComponents()
            return
        }
        
        var floorSwitchBuildings = [PWBuilding]()
        var floorSwitchFloors = [PWFloor]()
        
        if let campus = mapView.campus {
            if let buildings = campus.buildings {
                floorSwitchBuildings.append(contentsOf: buildings)
            }
        } else if let building = mapView.building {
            floorSwitchBuildings.append(building)
        }
        
        guard !floorSwitchBuildings.isEmpty else {
            floorSwitchContainerView.isHidden = true
            floorSwitchPickerView.reloadAllComponents()
            return
        }
        
        let buildingIndex = floorSwitchBuildings.firstIndex { (building) -> Bool in
            building.identifier == mapView.currentFloor?.building?.identifier
        }
        
        guard let currentBuildingIndex = buildingIndex else {
            floorSwitchContainerView.isHidden = true
            floorSwitchPickerView.reloadAllComponents()
            return
        }
        
        if let currentBuildingFloors = floorSwitchBuildings[currentBuildingIndex].floors {
            floorSwitchFloors.append(contentsOf: currentBuildingFloors)
        }
        
        let floorIndex = floorSwitchFloors.firstIndex { (floor) -> Bool in
            floor.floorID == mapView.currentFloor?.floorID
        }
        
        self.floorSwitchBuildings.append(contentsOf: floorSwitchBuildings)
        self.floorSwitchFloors.append(contentsOf: floorSwitchFloors)
        
        floorSwitchPickerView.reloadAllComponents()
        
        floorSwitchPickerView.selectRow(currentBuildingIndex, inComponent: Components.building, animated: false)
        
        if let currentFloorIndex = floorIndex {
            floorSwitchPickerView.selectRow(currentFloorIndex, inComponent: Components.floor, animated: false)
        }
        
        floorSwitchContainerView.isHidden = false
    }
}

// MARK: - UIPickerViewDelegate

extension LocationSharingViewController: UIPickerViewDelegate {
    private var buildingComponentWidth: CGFloat {
        return floorSwitchPickerView.frame.width / 2 // * (1 / 3.0)
    }
    
    private var floorComponentWidth: CGFloat {
        return floorSwitchPickerView.frame.width / 2 // * (2 / 3.0)
    }
    
    private var componentHeight: CGFloat {
        return 40
    }
    
    private var rowWidthMargin: CGFloat {
        return 30
    }
    
    func pickerView(_ pickerView: UIPickerView, widthForComponent component: Int) -> CGFloat {
        return component == Components.building ? buildingComponentWidth : floorComponentWidth
    }
    
    func pickerView(_ pickerView: UIPickerView, rowHeightForComponent component: Int) -> CGFloat {
        return componentHeight
    }
    
    func pickerView(_ pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusing view: UIView?) -> UIView {
        let label: UILabel
        
        if let reusableLabel = view as? UILabel {
            label = reusableLabel
        } else {
            let frame: CGRect
            let textAlignment: NSTextAlignment
            
            if component == Components.building {
                frame = .init(origin: .zero, size: .init(width: buildingComponentWidth - rowWidthMargin, height: componentHeight))
                textAlignment = .right
            } else if component == Components.floor {
                frame = .init(origin: .zero, size: .init(width: floorComponentWidth - rowWidthMargin, height: componentHeight))
                textAlignment = .left
            } else {
                frame = .zero
                textAlignment = .left
            }
            
            label = UILabel(frame: frame)
            label.textColor = .black
            label.textAlignment = textAlignment
            label.numberOfLines = 2
            label.font = .systemFont(ofSize: 13, weight: .medium)
        }
        
        if component == Components.building {
            label.text = floorSwitchBuildings[row].name
        } else if component == Components.floor {
            label.text = floorSwitchFloors[row].name
        } else {
            label.text = nil
        }
        
        return label
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        if component == Components.building {
            floorSwitchFloors.removeAll()
            
            guard let floors = floorSwitchBuildings[row].floors, let firstFloor = floors.first else {
                pickerView.reloadComponent(Components.floor)
                return
            }
            
            floorSwitchFloors.append(contentsOf: floors)
            
            pickerView.reloadComponent(Components.floor)
            
            pickerView.selectRow(0, inComponent: Components.floor, animated: false)
            
            mapView.currentFloor = firstFloor
        } else if component == Components.floor {
            mapView.currentFloor = floorSwitchFloors[row]
        }
    }
}

// MARK: - UIPickerViewDataSource

extension LocationSharingViewController: UIPickerViewDataSource {
    private enum Components {
        static let building = 0
        static let floor = 1
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 2
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        let numberOfRows: Int
        
        if component == Components.building {
            numberOfRows = floorSwitchBuildings.count
        } else if component == Components.floor {
            numberOfRows = floorSwitchFloors.count
        } else {
            numberOfRows = 0
        }
        
        return numberOfRows
    }
}
